sql: Add support for altering sequence options identity columns

delphamk opened the pull request:
Informs: #110010

This change introduces the 'ALTER [COLUMN] column_name SET sequence_option | RESTART [ [ WITH ] restart ] }' command. This enhancement allows for the modification of identity column sequence options, thereby providing more precise control over the behavior of identity columns.

Release note (sql change): Identity columns now support enhanced sequence management through the 'ALTER [COLUMN] column_name SET sequence_option' and 'ALTER [COLUMN] column_name RESTART [WITH restart]' commands. This update facilitates the fine-tuning of identity column sequences.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119432)
<!-- Reviewable:end -->

andrew-delph commented:
Here is a screenshot of the new parser diagrams:

![image](https://github.com/cockroachdb/cockroach/assets/8856279/a4ed09dc-22b1-4373-8f84-c9b6937d4556)
![image](https://github.com/cockroachdb/cockroach/assets/8856279/d89fcb03-7bb9-4f2d-b56c-8750446b785a)

andrew-delph commented on pkg/sql/alter_table.go:
```diff
@@ -1057,6 +1058,43 @@ func applyColumnMutation(
+			f.Printf(" CACHE %d", opts.CacheSize)
+		}
+		s := f.CloseAndGetString()
+		col.ColumnDesc().GeneratedAsIdentitySequenceOption = &s
```
Is it possible that 'GeneratedAsIdentitySequenceOption' can be overridden by another transaction?
It will be parsed down stream when writing to the information_schema.

fqazi reviewed (changes requested):
Great work @andrew-delph and thanks for doing this. Its fairly close just some minor comments / nits.

> Reviewed 10 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119432#-:-NsoLCsh4Ror4GjzSmAc:ba15ktr)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrew-delph)

___
*[`pkg/sql/alter_table.go` line 1071 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119432#-NsoINGl2WXU0jh49yiz:-NsoINGm-FjDLKo-JPVm:b-jaajr) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f61ae5b03bbec235bdd07fafa59ba22d7a6f57/pkg/sql/alter_table.go#L1071)):*
> ```Go
> 		// It is assummed that an identiy column owns only one sequence.
> 		if col.NumUsesSequences() != 1 {
> 			return colinfo.NewUndefinedColumnError(string(col.ColName()))
> ```

I don't think this should happen so lets use errors.AssertionFailed if we see it.
___
*[`pkg/sql/alter_table.go` line 1088 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119432#-NsoKbb29Mvt_yarUERz:-NsoKbb29Mvt_yarUES-:b-j7qi2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f61ae5b03bbec235bdd07fafa59ba22d7a6f57/pkg/sql/alter_table.go#L1088)):*
> ```Go
> 		opts := seqDesc.GetSequenceOpts()
> 		f := tree.NewFmtCtx(tree.FmtSimple)
> 		f.Printf(" MINVALUE %d", opts.MinValue)
> ```

Might be cleaner to just make this out of tree.SequenceOption and then format it out
___
*[`pkg/sql/alter_table.go` line 1096 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119432#gh-1501870176:-NsoK6SpAS1fQNl3r0mP:b-crtbwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f61ae5b03bbec235bdd07fafa59ba22d7a6f57/pkg/sql/alter_table.go#L1096)):*
<details><summary><i>Previously, andrew-delph (Andrew Delph) wrote…</i></summary><blockquote>

Is it possible that 'GeneratedAsIdentitySequenceOption' can be overridden by another transaction?
It will be parsed down stream when writing to the information_schema.

</blockquote></details>

So other concurrent writes can happen, but your transaction will be forced to retry because of the conflict. So, theres no danger and this will work correctly.
___
*[`pkg/sql/logictest/testdata/logic_test/information_schema` line 2976 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119432#-NsoL56eEQnk49s-yhm7:-NsoL56eEQnk49s-yhm8:b-mzfhdd) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f61ae5b03bbec235bdd07fafa59ba22d7a6f57/pkg/sql/logictest/testdata/logic_test/information_schema#L2976)):*
> ```
> 
> statement ok
> CREATE DATABASE test_identity
> ```

You can just use the same database for this test, no need to make a new one


<!-- Sent from Reviewable.io -->

delphamk committed:
sql: Add support for altering sequence options identity columns

Informs: #110010

This change introduces the 'ALTER [COLUMN] column_name SET
sequence_option | RESTART [ [ WITH ] restart ] }' command.
This enhancement allows for the modification of identity column
sequence options, thereby providing more precise control over
the behavior of identity columns.

Release note (sql change): Identity columns now support enhanced
sequence management through the 'ALTER [COLUMN] column_name SET
sequence_option' and 'ALTER [COLUMN] column_name RESTART [WITH
restart]' commands. This update facilitates the fine-tuning of
identity column sequences.

andrew-delph commented:
Thank you @fqazi for the comments! 
I have pushed some changes now.

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 12 of 12 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119432#-:-NsuCtcy5JJyn6KWuJe0:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andrew-delph)



<!-- Sent from Reviewable.io -->

fqazi commented:
bors r+
